domain-models-runtime/src/main/java/org/folio/rest/persist/PostgresConnectionManager.java

```diff
@@ -0,0 +1,96 @@
+package org.folio.rest.persist;
+
+import io.vertx.pgclient.PgConnection;
+import io.vertx.sqlclient.Pool;
+import io.vertx.core.Future;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.folio.rest.tools.utils.Envs;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class PostgresConnectionManager {
+  private static final Logger LOG = LogManager.getLogger(PostgresConnectionManager.class);
+  private static final int DEFAULT_CACHE_SIZE = 4;
+  private final int cacheSizeLimit;
+  private final List<PgConnection> connectionCache = new ArrayList<>();
+  private String currentTenant = "";
+  private int currentIndex;
+
+  public PostgresConnectionManager () {
+    cacheSizeLimit = Envs.getEnv(Envs.DB_MAXSHAREDPOOLSIZE) != null ?
+        Integer.parseInt(Envs.getEnv(Envs.DB_MAXSHAREDPOOLSIZE)) : DEFAULT_CACHE_SIZE;
+  }
+
+  public void clearCache() {
+    currentTenant = "";
+    currentIndex = 0;
+    connectionCache.clear();
+
+    LOG.debug("Cleared connection manager");
+  }
+
+  public Future<PgConnection> getConnection(Pool pool, String schemaName, String tenantId) {
+    if (!PostgresClient.sharedPgPool) {
+      LOG.debug("Not in shared pool mode");
+      return pool.getConnection().map(PgConnection.class::cast);
+    }
+
+    return getCachedConnection(pool, schemaName, tenantId);
+  }
+
+  private Future<PgConnection> getCachedConnection(Pool pool, String schemaName, String tenantId) {
+    if (tenantId.equals(currentTenant)) {
```

Comment:
Not sure how this will play out when multiple tenants use the system actively and concurrently.

Reply:
The key to the idea working is that all callers have the same view of the cache and the current tenant id. If the tenant id happens to change, the cache is destroyed and we start building a new one. This of course it will perhaps multiple times a second, but sometimes not for longer periods. The hope is that in these longer periods, when requests flow in from a single tenant, we can reduce round trips during these batches.

Maybe the implementation needs work, but that is the design I'm going for. Everything static and shared with unified view of the current tenant and the cache that causes no conflicts and is fast.